### ISSUE TITLE
[Modules][Dashboard] The toast to reload the page due to the index pattern fields were refreshed appear constatly despite to refresh the page previously.

### ISSUE THREAD

**Issue by Desvelao** [wazuh/wazuh-kibana-app#3918]
| Wazuh | Elastic | Rev  | Security |
| ----- | ------- | ---- | -------- |
| 4.x   | 7.x     | 4xxx | Basic, ODFE, Xpack |

| Browser |
| ------- |
| Chrome, Firefox, Safari, etc|

**Description**
When accessing some dashboards of modules, a toast appears saying the index pattern fields were refreshed and it is required to reload the page. After reloading the page, the toast appears again.

**Steps to reproduce**
1. Navigate to `Modules/Google Cloud Platform/Dashboard`
2. See the toast
3. Refresh the page by clicking on the `Reload page` button or refreshing the page with the browser button or shortcut.
4. The toast appears again.

**Expected Result**
1. The toast should appear constantly.

**Screenshots**
![image](https://user-images.githubusercontent.com/34042064/160106845-96e10e71-cb30-42b6-9aa0-565f7cb5acff.png)

**Additional context**
Add any other context about the problem here. Here you can paste log entries too or any other useful information that may help with the issue.

**Comment by Desvelao** [comment]
# Research
When building a visualization requires a non-existent field in the index pattern, [an error is caught](https://github.com/wazuh/wazuh-kibana-app/blob/4.3-7.10/public/kibana-integrations/kibana-vis.js#L310-L336). At the current source code status, the field name is extracted from the error message and is checked if this exists in the [known fields](https://github.com/wazuh/wazuh-kibana-app/blob/4.3-7.10/public/utils/known-fields.js) and used to refresh the index pattern fields as an extra field with builtin metadata and the existent fields in the indices.

For some modules such as Office 365 or GitHub, the fields used in the visualizations are not included in the builtin known fields and this causes the field can't be found there and can't be added to the index pattern fields despite the field really doesn't exist in the indices. This causes when there is no data related to alerts of these modules, the toast message of `The index pattern was refreshed successfully` to be displayed every time despite refreshing the browser tab.

## Possible solutions
- Sure the fields used by the visualization are available in the indices before creating the visualizations. This approach changes the current logic.
- Sure all fields used in visualizations related to alerts are included in the [known fields](https://github.com/wazuh/wazuh-kibana-app/blob/4.3-7.10/public/utils/known-fields.js) fields. This approach could cause the toast appears again if the index pattern fields were refreshed by the plugin, when checking the fields in the indices or if the user uses the index pattern management to apply a manual refresh of fields (`Stack management/Index pattern`) doing the fields are removed.
- After catching the error of there a missing field, refresh all index patterns fields if the missing field exists in the indices.

**Comment by Desvelao** [comment]
## Possible solution
For the approach:
> After catching the error of a missing field, refresh all fields of the index pattern if the missing field exists in the indices.

I created the next patch that can be appliend in the current status of `4.3-7.10` branch:
```diff
diff --git a/public/components/visualize/wz-visualize.js b/public/components/visualize/wz-visualize.js
index 2b578f2a9..065aa5a06 100644
--- a/public/components/visualize/wz-visualize.js
+++ b/public/components/visualize/wz-visualize.js
@@ -122,20 +122,16 @@ export const WzVisualize = compose(
       this.setState({ expandedVis: this.state.expandedVis === id ? false : id });
     };
 
-    refreshKnownFields = async (newField = null) => {
-      if (newField && newField.name) {
-        this.newFields[newField.name] = newField;
-      }
+    refreshKnownFields = async () => {
       if (!this.state.hasRefreshedKnownFields) {
         // Known fields are refreshed only once per dashboard loading
         try {
           this.setState({ hasRefreshedKnownFields: true, isRefreshing: true });
           if(satisfyPluginPlatformVersion('<7.11')){
-            await PatternHandler.refreshIndexPattern(this.newFields);
+            await PatternHandler.refreshIndexPattern();
           };
           this.setState({ isRefreshing: false });
           this.reloadToast();
-          this.newFields = {};
         } catch (error) {
           this.setState({ isRefreshing: false });
           const options = {
diff --git a/public/kibana-integrations/kibana-vis.js b/public/kibana-integrations/kibana-vis.js
index ea6460099..4004b0688 100644
--- a/public/kibana-integrations/kibana-vis.js
+++ b/public/kibana-integrations/kibana-vis.js
@@ -18,7 +18,7 @@ import { connect } from 'react-redux';
 import { LoadedVisualizations } from '../factories/loaded-visualizations';
 import { RawVisualizations } from '../factories/raw-visualizations';
 import { VisHandlers } from '../factories/vis-handlers';
-import { AppState } from '../react-services';
+import { AppState, SavedObject } from '../react-services';
 import { TabVisualizations } from '../factories/tab-visualizations';
 import store from '../redux/store';
 import { updateMetric } from '../redux/actions/visualizationsActions';
@@ -312,7 +312,7 @@ class KibanaVis extends Component {
           this.tabVisualizations.addDeadVis();
           return this.renderComplete();
         }
-        const match = error.message.match(/id:(.*)\)/);
+        const match = error.message.match(/id: (.*)\)/);
         this.deadField = match[1] || true;
         if (this.props.refreshKnownFields && !this.hasRefreshed) {
           this.hasRefreshed = true;
@@ -322,11 +322,14 @@ class KibanaVis extends Component {
           this.renderInProgress = false;
           this.rendered = false;
 
-          // if there's a field name it looks for known fields structures
-          const foundField =
-            match[1] && KnownFields.find((field) => field.name === match[1].trim());
-
-          await this.props.refreshKnownFields(foundField);
+          const currentPattern = AppState.getCurrentPattern();
+          const pattern = await getDataPlugin().indexPatterns.get(currentPattern);
+          const fields = await SavedObject.getIndicesFields(pattern.title);
+          console.log({fields});
+          const foundField = fields.find(field => field.name === match[1]);
+          if(foundField){
+            await this.props.refreshKnownFields();
+          };
         }
         this.renderInProgress = false;
         return this.myRender(raw);

```

**Comment by Desvelao** [comment]
I noticed that the index pattern includes fields related to `aws`, for example, `data.aws.bytes`, despite I have no alerts related to `aws`. I was researching and this could be caused due to some or multiple indices that could have defined these fields in their mappings. These mappings are coming from the template added in Filebeat to use with the Wazuh.

So, I guess if we add the fields used by the visualizations in the affected modules, Office365 and GitHub, to the template used by Wazuh and this is set to Elasticsearch, the new indices will contain their fields in the mapping, and the endpoint `/api/index-pattern/_fields_for_widcard?pattern=<PATTERN_TITLE>` will return the fields despite the indices has no data related to these modules.

**Comment by Desvelao** [comment]
@asteriscos and I were discussing this issue and researching the options to solve.

Current behavior:
The visualizations in the `Modules/<Module>/Dashboard` are built when accessing the section. They use some expected fields to be present in the fields of the index pattern. If the field is not found, then an error is thrown and caught where the plugin gets the field name and uses a method managed by `WzVisualize` component to add the field to the fields of the index pattern. This method is executed only one time, despite there were more visualizations with missing fields.

Deseable:
Our intention was that each visualization with the error of a missing field, report the field to the parent component to centralize the logic to update the fields of index pattern taking in account all the fields that are missing. 

We discussed how this could be done and managed some options to centralize the missing fields in the `WzVisualize` component using the component state or using the Redux store. We review how the drilldown for the `GitHub` or `Office 365` modules are made and we saw that the `WzVisualize` component is not used because is used the component that build the individual visualization. This means that we would have to create similar logic for the two different uses cases (if there is not more cases). As this approach, increases the technical debt, we think that we should fix this problem if unify the behavior.

Current proposal:
We were discussing this with @gdiazlo and we agree to update the template used by Filebeat and the know fields that use the plugin to create the index pattern of alerts when there are no related indices. This is a partial solution that could avoid the toast appearing each time due to Kibana getting the fields of index patterns. Related to https://github.com/wazuh/wazuh-kibana-app/issues/3918#issuecomment-1080621849.
